keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ResourceService.kt

```diff
@@ -0,0 +1,192 @@
+    // this so we don't inadvertently start actuating on a resource that had been previously paused,
+    // without explicit action from the user to resume.
+    if (actuationPauser.isPaused(id)) {
+      return PAUSED
+    }
+
+    val history = getEnrichedEventHistory(id, 10)
+    return when {
+      history.isHappy() -> HAPPY
+      history.isUnhappy() -> UNHAPPY
+      history.isDiff() -> DIFF
+      history.isActuating() -> ACTUATING
+      history.isError() -> ERROR
+      history.isCreated() -> CREATED
+      history.isVetoed() -> VETOED
+      history.isResumed() -> RESUMED
+      history.isCurrentlyUnresolvable() -> CURRENTLY_UNRESOLVABLE
+      else -> UNKNOWN
+    }
+  }
+
+  private fun List<PersistentEvent>.isHappy(): Boolean {
+    return first() is ResourceValid || first() is ResourceDeltaResolved
+  }
+
+  private fun List<PersistentEvent>.isActuating(): Boolean {
+    return first() is ResourceActuationLaunched || first() is ResourceTaskSucceeded ||
+      // we might want to move ResourceTaskFailed to isError later on
+      first() is ResourceTaskFailed
+  }
+
+  private fun List<PersistentEvent>.isError(): Boolean {
+    return first() is ResourceCheckError
+  }
+
+  private fun List<PersistentEvent>.isCreated(): Boolean {
+    return first() is ResourceCreated
+  }
+
+  private fun List<PersistentEvent>.isDiff(): Boolean {
+    return first() is ResourceDeltaDetected || first() is ResourceMissing
+  }
+
+  private fun List<PersistentEvent>.isPaused(): Boolean {
+    val appPaused = firstOrNull { it is ApplicationActuationPaused }?.timestamp
+    val appResumed = firstOrNull { it is ApplicationActuationResumed }?.timestamp
+
+    return first() is ResourceActuationPaused ||
+      // If the app was paused and has not yet resumed
+      (appPaused != null && (appResumed == null || appResumed.isBefore(appPaused)))
+  }
+
+  private fun List<PersistentEvent>.isVetoed(): Boolean {
+    return first() is ResourceActuationVetoed
+  }
+
+  private fun List<PersistentEvent>.isResumed(): Boolean {
+    return first() is ResourceActuationResumed || first() is ApplicationActuationResumed
+  }
+
+  private fun List<PersistentEvent>.isCurrentlyUnresolvable(): Boolean {
+    return first() is ResourceCheckUnresolvable
+  }
+
+  /**
+   * Checks last 10 events for flapping between only ResourceActuationLaunched and ResourceDeltaDetected
+   */
+  private fun List<PersistentEvent>.isUnhappy(): Boolean {
+    val recentSliceOfHistory = this.subList(0, Math.min(10, this.size))
+    val filteredHistory = recentSliceOfHistory.filter { it is ResourceDeltaDetected || it is ResourceActuationLaunched }
+    if (filteredHistory.size != recentSliceOfHistory.size) {
+      // there are other events, we're not thrashing.
+      return false
+    }
+    return true
+  }
+
+  /**
+   * Returns a list of [ResourceSummary] for the specified application.
+   *
+   * This function assumes there's a single delivery config associated with the application.
```

Comment:
This comment is not needed, because that's an application-wide assumption

Reply:
Where is the assumption written, though? 🤔

Reply:
What do you mean? We have specific checks for it (like [here](https://github.com/spinnaker/keel/blob/master/keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepository.kt#L101)), and a unique index constraint on the application field of the delivery config table. If you mean is there text anywhere that says that, I'm not sure, but regardless this comment is not the place to note an application wide assumption